File: packages/ios-sdk/Sources/PhosraSDK/Enforcement/WebFilterEnforcer.swift
import Foundation

#if canImport(ManagedSettings)
import ManagedSettings
#endif

/// Enforces web content filtering using Apple's ManagedSettings framework.
///
/// Handles:
/// - Web content filter level (none / moderate / strict)
/// - Safe search enforcement
/// - Domain blocklist and allowlist
/// - Category-based blocking
///
/// - Important: ManagedSettings web content filtering works through
///   `ManagedSettingsStore.webContent`. The auto-filter provides Apple's built-in
///   web content classification. Custom domain lists supplement the auto-filter.
@available(iOS 16.0, macOS 13.0, *)
final class WebFilterEnforcer {
    #if canImport(ManagedSettings) && os(iOS)
    private let store = ManagedSettingsStore()
    #endif

    /// Apply web filter settings to the device.
    ///
    /// - Parameter webFilter: Web filter configuration from the compiled policy.
    /// - Returns: A category report describing the enforcement outcome.
    func apply(_ webFilter: WebFilter) -> EnforcementReport.CategoryReport {
        var report = EnforcementReport.CategoryReport()
        report.framework = "ManagedSettings"

        #if canImport(ManagedSettings) && os(iOS)
        do {
            var details: [String] = []

            // MARK: - Web Content Filter Level
            //
            // MARK: - TODO: Apply web content filter via ManagedSettingsStore.
            //
            // For "strict" mode:
            //   store.webContent.autoFilter = .strict
            // For "moderate" mode:
            //   store.webContent.autoFilter = .moderate
            // For "none":
            //   store.webContent.autoFilter = nil
            //
            // Note: autoFilter is a WebContentSettings.AutoFilter enum, not a string.
            // The exact API may vary by iOS version.

            switch webFilter.level {
            case "strict":
                details.append("Filter level: strict")
            case "moderate":
                details.append("Filter level: moderate")
            default:
                details.append("Filter level: none")
            }

            // MARK: - Safe Search
            if webFilter.safeSearch {
                // MARK: - TODO: Safe search enforcement.
                //
                // Apple's ManagedSettings does not have a direct safe-search toggle.
                // Options:
                // 1. Use WebContentFilter network extension to rewrite search URLs
                // 2. Block non-safe-search URLs at the DNS level
                // 3. Rely on the auto-filter which includes some safe-search behavior
                details.append("Safe search: enabled")
            }

            // MARK: - Blocked Domains
            if !webFilter.blockedDomains.isEmpty {
                // MARK: - TODO: Apply blocked domains via ManagedSettingsStore.
                //
                // Use store.webContent.blockedByFilter to add URLs:
                //   let urls = webFilter.blockedDomains.compactMap { URL(string: "https://\($0)") }
                //   store.webContent.blockedByFilter = WebContentSettings.FilterOverride(urls: urls)
                details.append("Blocked domains: \(webFilter.blockedDomains.count)")
            }

            // MARK: - Allowed Domains
            if !webFilter.allowedDomains.isEmpty {
                // MARK: - TODO: Apply allowed domains via ManagedSettingsStore.
                //
                // Use store.webContent.allowedByFilter to add exception URLs.
                details.append("Allowed domains: \(webFilter.allowedDomains.count)")
            }

            // MARK: - Blocked Categories
            if !webFilter.blockedCategories.isEmpty {
                // MARK: - TODO: Map Phosra web categories to Apple's content filter categories.
                //
                // Apple's auto-filter handles standard categories. Custom categories
                // (e.g., "gambling", "dating") may need to be enforced via domain blocklists
                // from the ApplePlatformMappings.
                details.append("Blocked categories: \(webFilter.blockedCategories.joined(separator: ", "))")
            }

            report.status = .partial
            report.details = details.joined(separator: "; ")
        }
        #else
        report.status = .unsupported
        report.details = "ManagedSettings is only available on iOS 16+"
        #endif

        return report
    }

    /// Remove all web filter restrictions.
    func remove() {
        #if canImport(ManagedSettings) && os(iOS)
        store.webContent = nil
        #endif
    }
}

File: packages/android-sdk/src/main/kotlin/com/phosra/sdk/enforcement/WebFilterEnforcer.kt
package com.phosra.sdk.enforcement

import android.content.Context
import com.phosra.sdk.models.WebFilter
import com.phosra.sdk.services.PhosraVpnService

/**
 * Enforces web content filtering using a local VPN service for DNS-level blocking.
 *
 * Strategy:
 * - Starts [PhosraVpnService] as a local VPN tunnel
 * - Intercepts DNS queries and blocks resolution for filtered domains
 * - Supports domain blocklists, allowlists, and category-based filtering
 * - Safe search enforcement redirects search engine queries through safe search
 *
 * This does NOT proxy all traffic. It operates at the DNS level only,
 * which is lightweight and does not impact connection speed for allowed sites.
 *
 * Required permissions:
 * - BIND_VPN_SERVICE (user must approve the VPN connection dialog)
 *
 * TODO: Full VPN-based DNS filtering requires extensive device testing.
 *       Consider using DNS-over-HTTPS provider APIs as an alternative on
 *       Android 9+ (Private DNS settings).
 */
class WebFilterEnforcer(private val context: Context) {

    private var currentFilter: WebFilter? = null
    private var isActive = false

    /**
     * Apply web filtering rules.
     *
     * @param webFilter The web filter configuration from the compiled policy.
     * @return A [CategoryReport] describing the enforcement result.
     */
    fun apply(webFilter: WebFilter): CategoryReport {
        currentFilter = webFilter

        // Check if there's anything to enforce
        val hasRules = webFilter.safeSearch ||
            webFilter.blockedDomains.isNotEmpty() ||
            webFilter.allowedDomains.isNotEmpty() ||
            webFilter.blockedCategories.isNotEmpty() ||
            webFilter.level.isNotBlank()

        if (!hasRules) {
            return CategoryReport(
                status = EnforcementStatus.ENFORCED,
                framework = "VpnService",
                details = "No web filter rules configured; VPN not started"
            )
        }

        // Start the VPN service for DNS filtering
        return try {
            PhosraVpnService.start(context, webFilter)
            isActive = true

            val details = buildString {
                append("Web filter active (level: ${webFilter.level.ifBlank { "custom" }}). ")
                if (webFilter.safeSearch) append("Safe search enforced. ")
                if (webFilter.blockedDomains.isNotEmpty()) {
                    append("${webFilter.blockedDomains.size} domains blocked. ")
                }
                if (webFilter.allowedDomains.isNotEmpty()) {
                    append("${webFilter.allowedDomains.size} domains allowed. ")
                }
                if (webFilter.blockedCategories.isNotEmpty()) {
                    append("${webFilter.blockedCategories.size} categories blocked.")
                }
            }

            CategoryReport(
                status = EnforcementStatus.ENFORCED,
                framework = "VpnService (DNS filtering)",
                details = details.trim()
            )
        } catch (e: Exception) {
            CategoryReport(
                status = EnforcementStatus.FAILED,
                framework = "VpnService",
                details = "Failed to start VPN: ${e.message}",
                error = e
            )
        }
    }

    /**
     * Remove web filtering and stop the VPN service.
     */
    fun remove() {
        PhosraVpnService.stop(context)
        currentFilter = null
        isActive = false
    }

    /**
     * Get the current enforcement status.
     */
    fun currentStatus(): CategoryReport {
        return if (currentFilter != null && isActive) {
            CategoryReport(
                status = EnforcementStatus.ENFORCED,
                framework = "VpnService (DNS filtering)",
                details = "Web filter active"
            )
        } else {
            CategoryReport(
                status = EnforcementStatus.PENDING,
                framework = "VpnService",
                details = "Web filter not active"
            )
        }
    }

    /**
     * Check if a domain is blocked by the current filter configuration.
     *
     * @param domain The domain to check (e.g., "example.com").
     * @return true if the domain should be blocked.
     */
    fun isDomainBlocked(domain: String): Boolean {
        val filter = currentFilter ?: return false
        val normalizedDomain = domain.lowercase().trimEnd('.')

        // If we have an allowlist, only allow those domains
        if (filter.allowedDomains.isNotEmpty()) {
            return !filter.allowedDomains.any { allowed ->
                normalizedDomain == allowed.lowercase() ||
                    normalizedDomain.endsWith(".${allowed.lowercase()}")
            }
        }

        // Check explicit blocklist
        if (filter.blockedDomains.any { blocked ->
                normalizedDomain == blocked.lowercase() ||
                    normalizedDomain.endsWith(".${blocked.lowercase()}")
            }) {
            return true
        }

        // Category-based blocking would require a domain categorization database
        // TODO: Integrate with a domain categorization API or local database
        //       for blocked_categories enforcement

        return false
    }

    /**
     * Get the safe search enforcement settings for major search engines.
     * Returns DNS rewrites or query parameter modifications needed.
     */
    fun getSafeSearchConfig(): Map<String, String> {
        if (currentFilter?.safeSearch != true) return emptyMap()

        // Safe search enforcement via DNS rewriting:
        // - Google: Redirect to forcesafesearch.google.com
        // - Bing: Redirect to strict.bing.com
        // - DuckDuckGo: Append safe=1 parameter
        // - YouTube: Redirect to restrictmoderate mode via DNS
        return mapOf(
            "www.google.com" to "forcesafesearch.google.com",
            "google.com" to "forcesafesearch.google.com",
            "www.bing.com" to "strict.bing.com",
            "bing.com" to "strict.bing.com",
            "www.youtube.com" to "restrict.youtube.com",
            "youtube.com" to "restrict.youtube.com",
            "m.youtube.com" to "restrict.youtube.com"
        )
    }
}
